app/src/main/java/org/simple/clinic/sync/indicator/SyncIndicatorStatusCalculator.kt

```diff
@@ -4,35 +4,39 @@ import com.f2prateek.rx.preferences2.Preference
 import io.reactivex.schedulers.Schedulers.io
 import org.simple.clinic.di.AppScope
 import org.simple.clinic.sync.DataSync
+import org.simple.clinic.sync.LastSyncedState
 import org.simple.clinic.sync.SyncGroup
 import org.simple.clinic.sync.SyncProgress
 import org.simple.clinic.util.Just
-import org.simple.clinic.util.Optional
 import org.threeten.bp.Clock
 import org.threeten.bp.Instant
 import javax.inject.Inject
-import javax.inject.Named
 
 @AppScope
 class SyncIndicatorStatusCalculator @Inject constructor(
     dataSync: DataSync,
     clock: Clock,
-    @Named("last_frequent_sync_succeeded_timestamp") private val lastSyncSuccessTimestamp: Preference<Optional<Instant>>,
-    @Named("last_frequent_sync_result") private val lastSyncProgress: Preference<Optional<SyncProgress>>
+    private val lastSyncProgress: Preference<LastSyncedState>
 ) {
   init {
     saveSyncResults(dataSync, clock)
   }
 
   private fun saveSyncResults(dataSync: DataSync, clock: Clock) {
-    dataSync
+    val syncResultsStream = dataSync
         .streamSyncResults()
         .distinctUntilChanged()
+
+    syncResultsStream
         .subscribeOn(io())
         .filter { (syncGroup, _) -> syncGroup == SyncGroup.FREQUENT }
-        .doOnNext { (_, result) -> lastSyncProgress.set(Just(result)) }
-        .filter { (_, result) -> result == SyncProgress.SUCCESS }
-        .map { lastSyncSuccessTimestamp.set(Just(Instant.now(clock))) }
+        .doOnNext { (_, results) ->
+            var updatedState = lastSyncProgress.get().copy(lastSyncProgress = Just(results))
+            if (results == SyncProgress.SUCCESS) {
+              updatedState = updatedState.copy(lastSyncSuccessTimestamp = Just(Instant.now(clock)))
+            }
```

Comment:
I really prefer avoiding variables. It'll be nice if we can do `val` everywhere. 

```kotlin
val updatedState = when (progress) {
  SUCCESS ->
    lastSyncProgress.get().copy(
        lastSyncProgress = Just(progress),
        lastSyncSuccessTimestamp = Just(Instant.now(clock)))

  SYNCING, FAILURE ->
    lastSyncProgress.get().copy(lastSyncProgress = Just(progress))
}
```

Using an exhaustive `when` will also give us safety to enum changes.